sql-plugin/src/main/scala/com/nvidia/spark/rapids/ScalableTaskCompletion.scala

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.util
+
+import org.apache.spark.TaskContext
+
+/**
+ * A handle that can be used to remove a callback if needed
+ */
+trait TaskCompletionCallbackHandle {
+  /**
+   * Removes the callback but does not call it if it has not already been called.
+   */
+  def removeCallback(): Unit
+
+  /**
+   * Removes the callback and calls it if it has not been called already.
+   */
+  def removeAndCall(): Unit
+}
+
+/**
+ * Provides a task completion listeners in Spark that you can remove if needed to help with scaling.
+ * Spark guarantees LIFO order to the callbacks, but we do not. If you need that kind of a
+ * guarantee then use the Spark task APIs directly.
+ */
+object ScalableTaskCompletion {
+  // Note: There are three levels of synchronization that can happen here. The order that they are
+  // grabbed should always be
+  //
+  // ScalableTaskCompletion -> TopLevelTaskCompletion -> UserTaskCompletion
+  //
+  // this is to avoid any potential deadlocks
+
+  /**
+   * Wrapper around a user callback function.
+   */
+  private class UserTaskCompletion(val tc: TaskContext,
+      f: Either[() => Unit, TaskContext => Unit])
+      extends TaskCompletionCallbackHandle
+          with Function[TaskContext, Unit] {
+    private var wasCalled = false
+
+    override def removeCallback(): Unit = {
+      val topLevel = ScalableTaskCompletion.getTopLevel(tc.taskAttemptId())
```

Comment:
why not call `remove` on the hashmap blindly, rather than get and remove?

Reply:
This API is removing a callback from a task. We get the collection object that holds task callbacks and check to see if it (the collection object) is null. If it is not we then remove ourselves, the callback, from that collection. I guess each handle could hold onto a reference to the original collection associated with it instead of asking the singleton for it.  Is that what you are asking?